technologydata/currency_utils.py

```diff
@@ -0,0 +1,461 @@
+
+    >>> CurrencyUtils.convert_and_adjust_currency(
+    ...     base_year_val=2020,
+    ...     deflator_function_name="example_deflator",
+    ...     target_currency="USD",
+    ...     pydeflate_path=pathlib.Path("/path/to/deflator"),
+    ...     data=pd.DataFrame({"unit": ["USD-2020", "EUR-2020"], "value": [100, 200], "region": ["US", "EU"]})
+    ... )
+    ...
+    # Returns a DataFrame with adjusted currency values.
+
+    """
+
+    @staticmethod
+    def ensure_currency_unit(
+        input_string: str, expected_format: str = r"[A-Z]{3}-\d{4}"
+    ) -> str | None:
+        r"""
+        Check if the input string contains a currency unit.
+
+        The method searches for a substring that matches the expected format and extracts it if found.
+
+        Parameters
+        ----------
+        input_string : str
+            The string to check.
+        expected_format : str
+            The string with the expected format.
+
+        Returns
+        -------
+        str
+            The matched currency unit if found, None otherwise.
+
+        Raises
+        ------
+        ValueError
+            If the input_string or the expected_format are not a string.
+
+        Examples
+        --------
+        >>> CurrencyUtils.ensure_currency_unit("The price is USD-2025", r"[A-Z]{3}-\d{4}")
+        'USD-2025'
+        >>> CurrencyUtils.ensure_currency_unit("No currency here", r"[A-Z]{3}-\d{4}")
+        None
+
+        """
+        if not isinstance(input_string, str) or not isinstance(expected_format, str):
+            raise ValueError("Input must be a string.")
+
+        currency_unit_pattern = re.compile(expected_format)
+        match = currency_unit_pattern.search(input_string)
+        return match.group(0) if match else None
+
+    @staticmethod
+    def get_deflate_row_function(
+        base_year: int,
+        deflator_name: str,
+        target_currency: str,
+        year: str,
+        iso_code: str,
+        target_value_column: str,
+    ):
+        deflation_function = deflation_function_registry.get(deflator_name)
+        if deflation_function is None:
+            raise ValueError(
+                f"Deflator function '{deflator_name}' not found in registry"
+            )
+
+        def deflate_row(row):
+            row_df = pd.DataFrame([row])
+            deflated_df = deflation_function(
+                data=row_df,
+                base_year=base_year,
+                source_currency=row[iso_code],
+                target_currency=target_currency,
+                id_column=iso_code,
+                year_column=year,
+                value_column="value",
+                target_value_column=target_value_column,
+            )
+            return deflated_df.iloc[0]
+
+        return deflate_row
+
+    @staticmethod
+    def convert_and_adjust_currency(
+        base_year_val: int,
+        deflator_function_name: str,
+        target_currency_country_code: str,
+        pydeflate_path: pathlib.Path,
```

Comment:
The user shouldn't need to worry about the `pydeflate_path` can we have a sensible default, e.g. simply let `pydeflate` manage it and only have this optional?